File: .agents/skills/discord-bot-engineer/assets/patterns/token-manager.mjs
/**
 * High-performance token authentication, token pool, and OAuth2 management for Discord bots (Node.js / ESM).
 */

const DISCORD_API_BASE = 'https://discord.com/api/v10';

export class TokenVerificationError extends Error {
  constructor(message, status = null, responseBody = null) {
    super(message);
    this.name = 'TokenVerificationError';
    this.status = status;
    this.responseBody = responseBody;
  }
}

export class DiscordTokenValidator {
  /**
   * Fast verification using Discord REST API directly (no gateway overhead).
   * @param {string} token
   * @param {'Bot' | 'Bearer'} tokenType
   * @returns {Promise<object>} User/Bot object
   */
  static async request(endpoint, token, tokenType = 'Bot') {
    const res = await fetch(`${DISCORD_API_BASE}${endpoint}`, {
      method: 'GET',
      headers: {
        Authorization: `${tokenType} ${token.trim()}`,
        'User-Agent': 'DiscordBotEngine/1.0',
      },
    });

    if (!res.ok) {
      const errText = await res.text();
      throw new TokenVerificationError(
        `Discord API error ${res.status}: ${res.statusText}`,
        res.status,
        errText
      );
    }

    return await res.json();
  }

  static async verifyBotToken(token) {
    return this.request('/users/@me', token, 'Bot');
  }

  static async verifyBearerToken(token) {
    return this.request('/users/@me', token, 'Bearer');
  }
}

export class DiscordOAuth2Manager {
  constructor({ clientId, clientSecret, redirectUri }) {
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.redirectUri = redirectUri;
  }

  async #postToken(params) {
    const body = new URLSearchParams(params);
    const res = await fetch(`${DISCORD_API_BASE}/oauth2/token`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'User-Agent': 'DiscordBotEngine/1.0',
      },
      body: body.toString(),
    });

    if (!res.ok) {
      const errText = await res.text();
      throw new TokenVerificationError(`OAuth2 Token error: ${res.status}`, res.status, errText);
    }

    return await res.json();
  }

  async exchangeCode(code) {
    return this.#postToken({
      client_id: this.clientId,
      client_secret: this.clientSecret,
      grant_type: 'authorization_code',
      code,
      redirect_uri: this.redirectUri,
    });
  }

  async refreshAccessToken(refreshToken) {
    return this.#postToken({
      client_id: this.clientId,
      client_secret: this.clientSecret,
      grant_type: 'refresh_token',
      refresh_token: refreshToken,
    });
  }

  async fetchUserInfo(accessToken) {
    return DiscordTokenValidator.verifyBearerToken(accessToken);
  }

  async fetchUserGuilds(accessToken) {
    const res = await fetch(`${DISCORD_API_BASE}/users/@me/guilds`, {
      headers: {
        Authorization: `Bearer ${accessToken.trim()}`,
        'User-Agent': 'DiscordBotEngine/1.0',
      },
    });
    if (!res.ok) {
      throw new TokenVerificationError(`Guild fetch failed: ${res.status}`, res.status, await res.text());
    }
    return await res.json();
  }
}

export class TokenPool {
  /**
   * @param {string[]} tokens
   */
  constructor(tokens = []) {
    this.tokens = tokens.map((t) => t.trim()).filter(Boolean);
    this.index = 0;
    this.backoffs = new Map();
  }

  addToken(token) {
    const clean = token.trim();
    if (clean && !this.tokens.includes(clean)) {
      this.tokens.push(clean);
    }
  }

  markRateLimited(token, retryAfterMs) {
    this.backoffs.set(token, Date.now() + retryAfterMs);
  }

  async getNextAvailableToken() {
    if (this.tokens.length === 0) {
      throw new Error('TokenPool is empty.');
    }

    const now = Date.now();
    for (let i = 0; i < this.tokens.length; i++) {
      const candidate = this.tokens[this.index % this.tokens.length];
      this.index++;
      const backoffUntil = this.backoffs.get(candidate) || 0;
      if (now >= backoffUntil) {
        return candidate;
      }
    }

    // All currently rate-limited; wait for earliest release
    const minWait = Math.min(...Array.from(this.backoffs.values())) - now;
    if (minWait > 0) {
      await new Promise((resolve) => setTimeout(resolve, minWait));
    }
    return this.tokens[this.index % this.tokens.length];
  }
}

File: .agents/skills/discord-bot-engineer/assets/patterns/token_manager.py
"""High-performance token authentication, token pool, and OAuth2 management for Discord bots.

Supports:
- Zero-gateway token verification & identity extraction
- Token Pool with health tracking and rotation
- Automatic Discord 429 rate limit backoff
- OAuth2 authorization code exchange & token refresh
- Multi-client bot login and lifecycle orchestration
"""

from __future__ import annotations

import asyncio
import itertools
import json
import logging
import time
from typing import Any, Dict, List, Optional
import urllib.parse
import urllib.request
import urllib.error

logger = logging.getLogger("discord_token_manager")
DISCORD_API_BASE = "https://discord.com/api/v10"


class TokenVerificationError(Exception):
    """Raised when token authentication fails."""


class DiscordTokenValidator:
    """Fast, lightweight token validator using direct REST API requests."""

    @staticmethod
    def _request(
        endpoint: str,
        token: str,
        token_type: str = "Bot",
        timeout: float = 10.0,
    ) -> Dict[str, Any]:
        url = f"{DISCORD_API_BASE}{endpoint}"
        req = urllib.request.Request(
            url,
            headers={
                "Authorization": f"{token_type} {token.strip()}",
                "User-Agent": "DiscordBotEngine/1.0",
            },
            method="GET",
        )
        try:
            with urllib.request.urlopen(req, timeout=timeout) as response:
                status = response.getcode()
                body = response.read().decode("utf-8")
                return json.loads(body)
        except urllib.error.HTTPError as err:
            err_body = err.read().decode("utf-8", errors="replace")
            raise TokenVerificationError(
                f"HTTP {err.code}: {err.reason} - {err_body}"
            ) from err
        except Exception as err:
            raise TokenVerificationError(f"Connection failed: {err}") from err

    @classmethod
    def verify_bot_token(cls, token: str, timeout: float = 10.0) -> Dict[str, Any]:
        """Verify bot token and return bot user object (id, username, avatar, flags)."""
        return cls._request("/users/@me", token, token_type="Bot", timeout=timeout)

    @classmethod
    def verify_bearer_token(cls, token: str, timeout: float = 10.0) -> Dict[str, Any]:
        """Verify OAuth2 Bearer token and return user profile."""
        return cls._request("/users/@me", token, token_type="Bearer", timeout=timeout)


class DiscordOAuth2Manager:
    """Manages OAuth2 access tokens, refresh tokens, and user data retrieval."""

    def __init__(self, client_id: str, client_secret: str, redirect_uri: str):
        self.client_id = client_id
        self.client_secret = client_secret
        self.redirect_uri = redirect_uri

    def _post_token(self, data: Dict[str, str]) -> Dict[str, Any]:
        url = f"{DISCORD_API_BASE}/oauth2/token"
        encoded = urllib.parse.urlencode(data).encode("utf-8")
        req = urllib.request.Request(
            url,
            data=encoded,
            headers={
                "Content-Type": "application/x-www-form-urlencoded",
                "User-Agent": "DiscordBotEngine/1.0",
            },
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=15.0) as response:
                return json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as err:
            err_body = err.read().decode("utf-8", errors="replace")
            raise TokenVerificationError(f"OAuth token error {err.code}: {err_body}") from err

    def exchange_code(self, code: str) -> Dict[str, Any]:
        """Exchange authorization code for access_token, refresh_token, and scope."""
        return self._post_token({
            "client_id": self.client_id,
            "client_secret": self.client_secret,
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": self.redirect_uri,
        })

    def refresh_access_token(self, refresh_token: str) -> Dict[str, Any]:
        """Use refresh_token to acquire a fresh access_token."""
        return self._post_token({
            "client_id": self.client_id,
            "client_secret": self.client_secret,
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
        })

    def fetch_user_info(self, access_token: str) -> Dict[str, Any]:
        """Fetch user profile with OAuth2 Bearer token."""
        return DiscordTokenValidator.verify_bearer_token(access_token)

    def fetch_user_guilds(self, access_token: str) -> List[Dict[str, Any]]:
        """Fetch guilds that the authenticated user belongs to."""
        url = f"{DISCORD_API_BASE}/users/@me/guilds"
        req = urllib.request.Request(
            url,
            headers={
                "Authorization": f"Bearer {access_token.strip()}",
                "User-Agent": "DiscordBotEngine/1.0",
            },
            method="GET",
        )
        with urllib.request.urlopen(req, timeout=15.0) as response:
            return json.loads(response.read().decode("utf-8"))


class TokenPool:
    """High-throughput token pool with health checks, rate limit backoff, and round-robin."""

    def __init__(self, tokens: List[str]):
        self.tokens = [t.strip() for t in tokens if t.strip()]
        self._cycler = itertools.cycle(self.tokens) if self.tokens else None
        self._lock = asyncio.Lock()
        self._backoffs: Dict[str, float] = {}

    def add_token(self, token: str) -> None:
        token = token.strip()
        if token and token not in self.tokens:
            self.tokens.append(token)
            self._cycler = itertools.cycle(self.tokens)

    def mark_rate_limited(self, token: str, retry_after: float) -> None:
        """Mark a token as temporarily rate-limited with expiry timestamp."""
        self._backoffs[token] = time.time() + retry_after
        logger.warning("Token %s... rate limited for %.2fs", token[:10], retry_after)

    async def get_next_available_token(self) -> str:
        """Get next valid, non-rate-limited token using round-robin."""
        if not self.tokens:
            raise ValueError("TokenPool is empty.")

        async with self._lock:
            now = time.time()
            for _ in range(len(self.tokens)):
                candidate = next(self._cycler)
                backoff_until = self._backoffs.get(candidate, 0.0)
                if now >= backoff_until:
                    return candidate

            # All tokens are backoff-limited; find minimum wait time
            min_wait = min(self._backoffs.values()) - now
            if min_wait > 0:
                await asyncio.sleep(min_wait)
            return next(self._cycler)
